gowin/arch.cc

```diff
@@ -0,0 +1,1248 @@
+        break;
+    case ID_N200:
+    case ID_N210:
+    case ID_N220:
+    case ID_N230:
+    case ID_N240:
+    case ID_N250:
+    case ID_N260:
+    case ID_N270:
+    case ID_S200:
+    case ID_S210:
+    case ID_S220:
+    case ID_S230:
+    case ID_S240:
+    case ID_S250:
+    case ID_S260:
+    case ID_S270:
+    case ID_E200:
+    case ID_E210:
+    case ID_E220:
+    case ID_E230:
+    case ID_E240:
+    case ID_E250:
+    case ID_E260:
+    case ID_E270:
+    case ID_W200:
+    case ID_W210:
+    case ID_W220:
+    case ID_W230:
+    case ID_W240:
+    case ID_W250:
+    case ID_W260:
+    case ID_W270:
+        len = id_X2;
+        break;
+    case ID_N800:
+    case ID_N810:
+    case ID_N820:
+    case ID_N830:
+    case ID_S800:
+    case ID_S810:
+    case ID_S820:
+    case ID_S830:
+    case ID_E800:
+    case ID_E810:
+    case ID_E820:
+    case ID_E830:
+    case ID_W800:
+    case ID_W810:
+    case ID_W820:
+    case ID_W830:
+        len = id_X8;
+        break;
+    case ID_GT00:
+    case ID_GT10:
+        glbsrc = id_SPINE_TAP_PCLK;
+        break;
+    case ID_GBO0:
+    case ID_GBO1:
+        glbsrc = id_TAP_BRANCH_PCLK;
+        break;
+    case ID_GB00:
+    case ID_GB10:
+    case ID_GB20:
+    case ID_GB30:
+    case ID_GB40:
+    case ID_GB50:
+    case ID_GB60:
+    case ID_GB70:
+        glbsrc = id_BRANCH_PCLK;
+        break;
+    default:
+        if (wire.str(this).rfind("SPINE", 0) == 0) {
+            glbsrc = ID_CENT_SPINE_PCLK;
+        } else if (wire.str(this).rfind("UNK", 0) == 0) {
+            glbsrc = ID_PIO_CENT_PCLK;
+        }
+        break;
+    }
+    if (len != id("")) {
+        return delayLookup(speed->wire.timings.get(), speed->wire.num_timings, len);
+    } else if (glbsrc != id("")) {
+        return delayLookup(speed->glbsrc.timings.get(), speed->glbsrc.num_timings, glbsrc);
+    } else {
+        DelayInfo info;
+        info.maxFall = 0;
+        info.minFall = 0;
+        info.maxRaise = 0;
+        info.minRaise = 0;
+        return info;
+    }
+}
+
+void Arch::read_cst(std::istream &in) {
+    std::regex iobre = std::regex("IO_LOC +\"([^\"]+)\" +([^ ;]+);");
+    std::smatch match;
+    std::string line;
+    while (!in.eof()) {
+        std::getline(in, line);
+        if(!std::regex_match(line, match, iobre)) continue;
+        std::cout << match[1] << " " << match[2] << std::endl;
+        IdString net = id(match[1]);
+        IdString pinname = id(match[2]);
+        const PairPOD *belname = pairLookup(package->pins.get(), package->num_pins, pinname.index);
+        if ( belname == nullptr)
+            log_error("Pin %s not found", pinname.c_str(this));
+        //BelId bel = getBelByName(belname->src_id);
+        // for (auto cell : sorted(cells)) {
+        //     std::cout << cell.first.str(this) << std::endl;
+        // }
+        auto it = cells.find(net);
+        if (it == cells.end())
+            log_error("Cell %s not found", net.c_str(this));
```

Comment:
Unless the Gowin tools behave like this, this should be a `continue` rather than an error - it's common for board constraint files to declare all the pins but only a subset are used in the design